File: backend/routes/vector_library.py
# -*- coding: utf-8 -*-
"""
向量库管理 API（插件式）。
"""

import logging

from flask import Blueprint, jsonify, request

from capabilities.vector_retrieval import get_vector_retrieval_capability
from services.vector_library_service import VectorLibraryServiceError

logger = logging.getLogger(__name__)

vector_library_bp = Blueprint('vector_library', __name__)


@vector_library_bp.route('/file-status', methods=['GET'])
def file_status():
    try:
        data = get_vector_retrieval_capability().file_status()
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('获取文件索引状态失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/index-file', methods=['POST'])
def index_file_with_vectorizer():
    try:
        data = get_vector_retrieval_capability().index_file(request.get_json(silent=True) or {})
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('按向量化器索引文件失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/delete-file', methods=['POST'])
def delete_file():
    try:
        data = get_vector_retrieval_capability().delete_file(request.get_json(silent=True) or {})
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('删除文件失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/vectorizers', methods=['GET'])
def list_vectorizers():
    try:
        data = get_vector_retrieval_capability().list_vectorizers()
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('列表向量化器失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/vectorizers', methods=['POST'])
def add_vectorizer():
    try:
        data = get_vector_retrieval_capability().add_vectorizer(request.get_json(silent=True) or {})
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('新增向量化器失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/vectorizers/<key>/activate', methods=['POST'])
def activate_vectorizer(key: str):
    try:
        data = get_vector_retrieval_capability().activate_vectorizer(key)
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('激活向量化器失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/vectorizers/<key>/docs', methods=['GET'])
def list_docs_by_vectorizer(key: str):
    try:
        data = get_vector_retrieval_capability().list_docs_by_vectorizer(key, request.args.get('collection'))
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('按向量化器查文档失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/migrate', methods=['POST'])
def migrate():
    try:
        data = get_vector_retrieval_capability().migrate(request.get_json(silent=True) or {})
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('迁移失败')
        return jsonify({'success': False, 'message': str(error)}), 500


@vector_library_bp.route('/vectorizers/<key>', methods=['DELETE'])
def delete_vectorizer(key: str):
    try:
        data = get_vector_retrieval_capability().delete_vectorizer(key)
        return jsonify({'success': True, 'data': data})
    except VectorLibraryServiceError as error:
        return jsonify({'success': False, 'message': error.message}), error.status_code
    except Exception as error:
        logger.exception('删除向量化器失败')
        return jsonify({'success': False, 'message': str(error)}), 500

File: frontend-client/src/api/vectorLibrary.js
/**
 * 向量库管理 API
 * 封装文件管理、向量索引、向量化器和向量搜索相关接口
 */

async function parseResponse(response) {
  const result = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(result.detail || result.message || `请求失败: ${response.status}`);
  }
  return result;
}

async function request(path, options = {}) {
  const response = await fetch(path, {
    headers: {
      'Content-Type': 'application/json',
      ...(options.headers || {}),
    },
    ...options,
  });
  return parseResponse(response);
}

// ── 文件管理 ────────────────────────────────────────────────────────────────

/**
 * 列出已上传的文件
 * @param {string[]} [extensions] - 文件扩展名过滤
 * @param {string[]} [mimeTypes] - MIME 类型过滤
 */
export async function listFiles(extensions, mimeTypes) {
  const query = new URLSearchParams();
  if (extensions?.length) query.set('extensions', extensions.join(','));
  if (mimeTypes?.length) query.set('mime_types', mimeTypes.join(','));
  const suffix = query.toString() ? `?${query.toString()}` : '';
  return request(`/api/files${suffix}`);
}

/**
 * 上传文件（支持多文件）
 * @param {FormData} formData - 包含 files 字段的表单数据
 */
export async function uploadFiles(formData) {
  const response = await fetch('/api/files/upload', {
    method: 'POST',
    body: formData,
    // 不设置 Content-Type，让浏览器自动设置 multipart/form-data
  });
  return parseResponse(response);
}

/**
 * 删除文件
 * @param {string} fileId - 文件 ID
 */
export async function deleteFile(fileId) {
  return request(`/api/files/${encodeURIComponent(fileId)}`, {
    method: 'DELETE',
  });
}

// ── 向量索引管理 ─────────────────────────────────────────────────────────────

/**
 * 获取文件的向量索引状态
 */
export async function getFileStatus() {
  return request('/api/vector-library/file-status');
}

/**
 * 将文件加入向量索引
 * @param {Object} body - { file_id, vectorizer_key?, collection? }
 */
export async function indexFile(body) {
  return request('/api/vector-library/index-file', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

/**
 * 删除文件的向量索引
 * @param {Object} body - { file_id, collection? }
 */
export async function deleteFileIndex(body) {
  return request('/api/vector-library/delete-file', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

// ── 向量化器管理 ─────────────────────────────────────────────────────────────

/**
 * 列出所有向量化器
 */
export async function listVectorizers() {
  return request('/api/vector-library/vectorizers');
}

/**
 * 添加向量化器
 * @param {Object} body - { key, model, ... }
 */
export async function addVectorizer(body) {
  return request('/api/vector-library/vectorizers', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

/**
 * 激活指定向量化器
 * @param {string} key - 向量化器 key
 */
export async function activateVectorizer(key) {
  return request(`/api/vector-library/vectorizers/${encodeURIComponent(key)}/activate`, {
    method: 'POST',
    body: JSON.stringify({}),
  });
}

/**
 * 删除向量化器
 * @param {string} key - 向量化器 key
 */
export async function deleteVectorizer(key) {
  return request(`/api/vector-library/vectorizers/${encodeURIComponent(key)}`, {
    method: 'DELETE',
  });
}

/**
 * 将上传文件首次导入向量集合
 * @param {Object} body - { file_id, document_id?, collection_name?, metadata?, chunk_size?, overlap? }
 */
export async function ingestFileToCollection(body) {
  return request('/api/vector/index', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

// ── 向量搜索 ─────────────────────────────────────────────────────────────────

/**
 * 向量相似度搜索
 * @param {Object} body - { query, top_k?, collection?, vectorizer_key? }
 */
export async function searchVectors(body) {
  return request('/api/vector/search', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

/**
 * 获取向量库健康状态
 */
export async function getVectorHealth() {
  return request('/api/vector/health');
}

/**
 * 列出某向量化器下的文档
 * @param {string} key - 向量化器 key
 * @param {Object} [params] - 查询参数
 */
export async function listDocsByVectorizer(key, params = {}) {
  const q = new URLSearchParams(params).toString();
  const suffix = q ? `?${q}` : '';
  return request(`/api/vector-library/vectorizers/${encodeURIComponent(key)}/docs${suffix}`);
}

/**
 * 在向量化器之间迁移数据
 * @param {Object} body - { from_key, to_key, collection? }
 */
export async function migrateVectorizer(body) {
  return request('/api/vector-library/migrate', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export default {
  listFiles,
  uploadFiles,
  deleteFile,
  getFileStatus,
  indexFile,
  deleteFileIndex,
  listVectorizers,
  addVectorizer,
  activateVectorizer,
  deleteVectorizer,
  listDocsByVectorizer,
  migrateVectorizer,
  ingestFileToCollection,
  searchVectors,
  getVectorHealth,
};
